### DISCUSSION TITLE
Technology stack (Frontend)

### DISCUSSION THREAD

**Discussion by steffen** [rarely-org/rarely-mobile#4] [General]
The majority of the engagement for this platform is expected to happen on mobile.
In order to deliver the best mobile experience to users while balancing the engineering effort, the current proposed technology for the client side mobile app is [React Native](https://reactnative.dev/).
No complex user interface interactions are expected (for now) that would require native iOS (Object C or Swift) or Android (Java or Kotlin) development.

**Pros**
- React Native allows sharing the code base between Android and iOS.
- It's using web technology with which all of the current maintainer team members are familiar with
- Household names such as Shopify seem to continue to invest in React Native, see list: https://reactnative.dev/showcase

**Comment by NVerban** [comment]
What demographics/regions do you expect to be main target audience? Worth considering what devices etc are common. for example, assumption that everyone has a modern up to date smart phone, or perhaps older generation devices. Also assumption that there is reliable internet access etc. Native app seems logical, but is there a need to support many different types of older devices and if so would that add any significant scope? PWA for example (although I am not a huge advocate for PWA), may allow for more universal, accessible solution depending on factors.

**Reply by NVerban** [reply]
+1 for react native

**Reply by NVerban** [reply]
Accessibility probably also key - such as voice control, spoken content etc. Choice of technology should also factor in implementation/ compatibility of such features across different devices / platforms.

**Reply by steffen** [reply]
@NVerban Spot on points and questions! 💖

> What demographics/regions do you expect to be main target audience?

Generally the target audience are all demographics and regions, except for age, where kids are not a target audience I would say. (It's more for teens, adults, parents or caregivers of kids.) [This discussion](https://github.com/rarely-org/rarely/discussions/2) goes a bit into the target audience area, which I still need to follow-up on.

> for example, assumption that everyone has a modern up to date smart phone, or perhaps older generation devices.

Very good point. I was just researching if I could find any specific information about device limitation and support for React Native, but I wasn't able to find anything.
My wishful thinking is that devices for the last 5 years or so are performant and modern enough to support React Native applications out of the box and that these are used by the majority of the users. Maybe this is something we can see with anonymous telemetrics later on or something we can find more data on.

> Also assumption that there is reliable internet access etc.

Being resilient to slow and spotty internet is probably something that we should incorporate with each feature development.
Generally speaking, from my knowledge and experience from traveling pre-COVID, I've been often surprised by the great mobile internet coverage around the globe, but then, I probably only travelled to 0.00000000001% of the world. 🙂 Graceful error handling, progress indicators and sensible timeouts should definitely be considered. 👍 

> PWA for example (although I am not a huge advocate for PWA)

Yeah, not a huge advocate either. Something to consider if we think it makes sense.

> Accessibility probably also key - such as voice control, spoken content etc. Choice of technology should also factor in implementation/ compatibility of such features across different devices / platforms.

💯 My hope is that a native mobile app (through React Native) makes this easier than a web-based.
I find component framework like [NativeBase](https://nativebase.io/), with built-in accessibility, very interesting.
NativeBase seems to be developed by the same people that develop https://react-native-aria.geekyants.com/.
(I don't have experience with either.)

Thank you! 🙇‍♂️

**Reply by NVerban** [reply]
I think React Native is a really sensible choice. You can build app to support say the last 5 years of device / operating system versions, but the more that is supported, the more work is required with testing etc, but there are tools available to make this much easier, for example: https://www.browserstack.com/

As a community project, might be worth considering https://expo.dev/ which would make building / testing much much easier.

**Reply by steffen** [reply]
> I think React Native is a really sensible choice. You can build app to support say the last 5 years of device / operating system versions, but the more that is supported, the more work is required with testing etc, but there are tools available to make this much easier, for example: https://www.browserstack.com/

@NVerban Thank you for your comment! I'm hoping we can support the majority of the devices and users without a lot of additional work and testing. Any additional support will need to be triaged just as any other enhancement and change.

> As a community project, might be worth considering https://expo.dev/ which would make building / testing much much easier.

I have added a scaffold for the mobile app using expo yesterday: https://github.com/rarely-org/rarely/commit/601c8d7ccc827d38660b9c2270de1316bfd3d2ca ☺️

I have also just published a brief video showing how to get started with it (mostly in GitHub Codespaces, but it's similar locally): https://www.youtube.com/watch?v=EnLL7i33Y_4

Full transparency: I have been learning a bit of React Native with this awesome [online course](https://codewithmosh.com/p/the-ultimate-react-native-course) which is using expo as well.

**Reply by NVerban** [reply]
Awesome! a quality video! TIL about GitHub Codespaces. Looks like it will make things a lot easier for everyone who will be able to contribute 😄 Nice work.
